pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6278,4 +6253,786 @@ SEC("tc")
+}
+
+SEC("kretprobe/__cgroup_bpf_run_filter_skb")
+int BPF_KRETPROBE(ret_cgroup_bpf_run_filter_skb)
+{
+    // runs AFTER the CGROUP/SKB eBPF program
+
+    event_data_t data = {};
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    // Both ingress & egress (mostly ingress) might run from a kthread and still
+    // need processing. Instead of "should_trace", check if there is a current
+    // skb timestamp entry and delete it (the current skb timestamp would have
+    // been added by __cgroup_bpf_run_filter_skb entry probe IF the socket is
+    // being traced).
+    //
+    // if (!should_trace(&data))
+    // return 0;
+
+    // pick from entry from entrymap
+    net_host_tid_evt_t tid = {0};
+    tid.host_tid = data.context.task.host_tid;
+    tid.net_event_id = internal_event_cgroup_bpf_run_filter_skb;
+    struct entry *entry = bpf_map_lookup_elem(&entrymap, &tid);
+    if (!entry) // no entry == no tracing
+        return 0;
+
+    // pick args from entry point's entry
+    struct sock *sk = (void *) entry->args[0];
+    struct sk_buff *skb = (void *) entry->args[1];
+
+    // cleanup entrymap
+    bpf_map_delete_elem(&entrymap, &tid);
+
+    // use skb timestamp as the key for cgroup/skb program context
+    u64 skbts = BPF_CORE_READ(skb, tstamp);
+
+    // only continue if netctx exists
+    net_event_context_t *netctx = bpf_map_lookup_elem(&cgrpctxmap, &skbts);
+    if (!netctx)
+        return 0;
+
+    // TODO: something to be done here ? (maybe just delete and it is good)
+
+    // delete netctx after cgroup ebpf program runs
+    bpf_map_delete_elem(&cgrpctxmap, &skbts);
+
+    return 0;
+}
+
+//
+// Type definitions and prototypes for protocol parsing:
+//
+
+typedef union iphdrs_t {
+    struct iphdr iphdr;
+    struct ipv6hdr ipv6hdr;
+} iphdrs;
+
+typedef union protohdrs_t {
+    struct tcphdr tcphdr;
+    struct udphdr udphdr;
+    struct icmphdr icmphdr;
+    struct icmp6hdr icmp6hdr;
+} protohdrs;
+
+// clang-format off
+#define CGROUP_SKB_HANDLE(name)                                                \
+static __always_inline u32 cgroup_skb_handle_##name(                           \
+    struct __sk_buff *ctx,                                                     \
+    net_event_context_t *neteventctx,                                          \
+    iphdrs *iphdrs,                                                            \
+    protohdrs *protohdrs                                                       \
+)
+
+CGROUP_SKB_HANDLE(family);
+CGROUP_SKB_HANDLE(proto);
+CGROUP_SKB_HANDLE(proto_tcp);
+CGROUP_SKB_HANDLE(proto_tcp_dns);
+CGROUP_SKB_HANDLE(proto_udp);
+CGROUP_SKB_HANDLE(proto_udp_dns);
+CGROUP_SKB_HANDLE(proto_icmp);
+CGROUP_SKB_HANDLE(proto_icmpv6);
+// clang-format on
+
+//
+// Network event submission functions
+//
+
+// clang-format off
+#define CGROUP_SKB_SUBMIT(name)                                                \
+static __always_inline u32 cgroup_skb_submit_##name(                           \
```

Comment:
I'll probably just pack these 4 variables (used in many hanlers) into a struct and pass the struct alongside them (instead of using macros to keep code more readable).